### ISSUE TITLE
PhysicalBone2D does not work

### ISSUE THREAD

**Issue by renleidedadi** [godotengine/godot#78367]
### Godot version

v4.0.3.stable.official [5222a99f5]

### System information

Windows 10 - Godot v4.0.3.stable - Vulkan (Forward+) - NVIDIA GeForce GTX 1650 (nvidia, 31.0.15.1694) - Intel Core i5-10400F CPU @ 2.90GHz

### Issue description

i make a 2D ragdoll with 4.0.3, PhysicalBone2D doesn't work, but RigidBody2D work, how to use PhysicalBone2D?

in this video, left is PhysicalBone2D, right is RigidBody2D:

https://github.com/godotengine/godot/assets/67264008/1fd8fe7b-95f7-4300-854b-4c07bf2045ea

### Steps to reproduce

1. download the project.
2. run main_scene.
3. left is PhysicalBone2D, right is RigidBody2D.

### Minimal reproduction project

[PhysicalBone2D in 4.0.3.zip](https://github.com/godotengine/godot/files/11779281/PhysicalBone2D.in.4.0.3.zip)

**Comment by stilestr** [comment]
I don't think you set up the skeleton correctly. However, I tried my best to follow the documentation and it was too confusing to get this to work properly. 

After googling some, it didn't seem like anyone has ever used PhysicalBone2D to make a ragdoll, and the only posts I found were unanswered questions asking how to get it to work.

So if anyone knows how to actually get this to work and can set up a working 2D skeleton using PhysicalBone2D, I'd be happy to help write some documentation or a tutorial on how to get it to work.

**Comment by renleidedadi** [comment]
> I don't think you set up the skeleton correctly. However, I tried my best to follow the documentation and it was too confusing to get this to work properly, so I couldn't do it either.
> 
> After googling some, it didn't seem like anyone has ever used PhysicalBone2D to make a ragdoll, and the only posts I found were unanswered questions asking how to get it to work.
> 
> So if anyone knows how to actually get this to work and can set up a working 2D skeleton using PhysicalBone2D, I'd be happy to help write some documentation or a tutorial on how to get it to work.

yes, PhysicalBone2D is difficult to use without tutorial, i think this issue should also be add "documentation" label. another situation is PhysicalBone2D might has something bugs: #78479 (PhysicalBone2D is broken?), but i ignorant of C++, i can't fix it

**Comment by stilestr** [comment]
> > I don't think you set up the skeleton correctly. However, I tried my best to follow the documentation and it was too confusing to get this to work properly, so I couldn't do it either.
> > After googling some, it didn't seem like anyone has ever used PhysicalBone2D to make a ragdoll, and the only posts I found were unanswered questions asking how to get it to work.
> > So if anyone knows how to actually get this to work and can set up a working 2D skeleton using PhysicalBone2D, I'd be happy to help write some documentation or a tutorial on how to get it to work.
> 
> yes, PhysicalBone2D is difficult to use without tutorial, i think this issue should also be add "documentation" label. another situation is PhysicalBone2D might has something bugs: #78479 (PhysicalBone2D is broken?), but i ignorant of C++, i can't fix it

Yeah, I just made that bug report yesterday while trying to research this one, lol. It seems that PhysicalBone2D might just be a highly neglected (and highly broken?) feature. Would be great to see if anyone has ever gotten it to work properly, cause it's hard to find any examples online

**Comment by Xorblax** [comment]
I concur on PhysicalBone2D not working great. I would honestly hold off on using them at all, as they're causing my editor to crash on startup now. I think what I did was move my normal bones that are referenced by physical bones, to another skeleton, or something. Don't exactly remember. This is the error log 
![image](https://github.com/godotengine/godot/assets/98433469/221620cc-7db5-469c-87f3-439d84798922)
Wish I could just remove the thing from my player scene but thankfully I didn't have that much work lost

**Comment by StrikeForceZero** [comment]
They seem to have their physics integration disabled in their constructor for some reason (ever since they were first added). https://github.com/godotengine/godot/blob/851bc640ddf7225a4b52bb15cc50c249df119953/scene/2d/physical_bone_2d.cpp#L290C1-L293

```cpp
PhysicalBone2D::PhysicalBone2D() {
	// Stop the RigidBody from executing its force integration.
	PhysicsServer2D::get_singleton()->body_set_collision_layer(get_rid(), 0);
	PhysicsServer2D::get_singleton()->body_set_collision_mask(get_rid(), 0);
	PhysicsServer2D::get_singleton()->body_set_mode(get_rid(), PhysicsServer2D::BodyMode::BODY_MODE_STATIC);
        // ...
```

**Comment by Dioinecail** [comment]
Hello, i tried the workaround above but the PhysicalBone2D still doesn't seem to work
My setup is:
```
PhysicalBody2D
  Skeleton2D
    Bone2D
    PhysicalBone2D
```

``Skeleton2D`` contains ``SkeletonModification2DPhysicalBones`` with exactly 1 bone set in Joint 0 NodePath

``PhysicalBone2D`` targets ``Bone2D`` but the editor still spits an error about incorrect modification stack

**Comment by StrikeForceZero** [comment]
@Dioinecail 
So there might have been some details left out of my original comment. I just attempted to do a clean room working example using @renleidedadi 's minimal reproduction. I eventually remembered that the physical bones **needs to have a joint + collision shape** to really do anything. _It's unfortunately really tedious_, but if you want a [working example you can see the modified code here](https://github.com/StrikeForceZero/godot-PhysicalBone2D-Workaround/commit/89debbfb44aa4db05688464d55e074a134737eb2)

Make sure to [read the comments](https://github.com/StrikeForceZero/godot-PhysicalBone2D-Workaround/blob/89debbfb44aa4db05688464d55e074a134737eb2/main_scene.gd#L35-L36) in the `main_scene.gd`

[godot-PhysicalBone2D-Workaround-example.zip](https://github.com/godotengine/godot/files/12705024/godot-PhysicalBone2D-Workaround-example.zip)

**Comment by Dioinecail** [comment]
@StrikeForceZero 
i got to the joint part while looking through the C++ code but i couldn't figure out that you need to have a collision shape but i guess it's only natural since it inherits RIgidBody2D 

gonna try the "full" setup tomorrow :)

**Comment by Dioinecail** [comment]
yup, it works now

you were right, there is probably a bug with initialization or constructor

making bones do 
```
for bone in yourTargetBones:
  bone.freeze = true
  bone.freeze = false
```
was enough to make it work

The most confusing part is setting everything up correctly
you need to have
``Skeleton2D`` with ``ModificationStack`` that contains ``SkeletonModification2DPhysicalBones`` and ``Joint X Nodepath``'s target your bones (if you change your bone name, it breaks) 
``PhysicalBone2D`` needs to have a ``PinJoint2D`` that has ``RigidBody2D`` or ``PhysicalBone2D`` connected to another ``PhysicalBone2D`` in the inspector
``PhysicalBone2D`` needs to have a ``CollisionShape2D`` as child
``PhysicalBone2D`` needs to have a ``Bone 2d Nodepath`` selected and correct ``Bone 2d Index`` selected as well!

and my hierarchy looks like this
```
RigidBody2D
  Node2D (just a root for skeleton)
    Skeleton2D
      BodyBone
        TargetBone2D
      PhysicalBone2D
        PinJoint2D
        CollisionShape2D
```

Overall yeah I would advise to avoid using PhysicalBone2Ds until they are properly documented and the freezing bug is fixed :D

**Comment by lilykiwi** [comment]
While I'd like to see this feature fixed and documented, it seems as though #65768 has an explanation for why it's not been touched yet:

> The ModificationStack made a lot of APIs implemented in the Skeleton that should not be there, since it was implemented without going through a formal review by core maintainers due to it was GSOC project. That must be removed and reimplement outside of Skeleton. Also, the current ModificationStack is significantly broken and needs to be refactored and reimplemented as a Node. Unfortunately, this refactoring is taking a long time, sorry.

Additionally that PR included a fix such that "PhysicalBone2D no longer relies on SkeletonModification2DPhysicalBones to update bone positions", although this wasn't present in #65801, the PR that was actually merged.

Unfortunately this implies that the feature is currently in a broken state, with an "experimental feature" warning on the documentation being the only indicator of this. Sadly it looks like we're just going to need to wait until this feature is reworked and in a stable state.

**Comment by YuriSizov** [comment]
> Removing those calls to the physics server allows the bones to act like rigid bodies. Very curious to know the reasoning behind this.

Just FYI, the problem here is things getting out of sync.  We have a chain of inheritance as follows:

- `PhysicalBone2D` < `RigidBody2D` < `PhysicsBody2D` < `CollisionObject2D`

And we have the 2D physics server. In Godot when a node is based on a server, there is often a two-part approach to properties. They exist both on the node class, and must be set in the server counterpart. The body mode member is one such property.

Normally, to keep things in sync you'd use `CollisionObject2D::set_body_mode`. But for some reason the `PhysicalBone2D` node modifies the server-side value directly via `PhysicsServer2D::get_singleton()->body_set_mode`. This leads to a desync where the member inherited from `CollisionObject2D` believes that the node is in the rigid mode (the default for any `RigidBody2D` instance), but the server thinks the node is the static mode (because that's what the `PhysicalBone2D` constructor tells it).

Toggling the freeze property implicitly forces `CollisionObject2D::set_body_mode` to be called with different values to re-sync the node with the server, but the fix can be effectively boiled down to this:

```gdscript
extends Node2D

@onready var _skeleton = %Skeleton


func _ready()->void:
	var modification_stack: SkeletonModificationStack2D = _skeleton.get_modification_stack()
	var modification_physical_bones: SkeletonModification2DPhysicalBones = modification_stack.get_modification(0)
	modification_physical_bones.fetch_physical_bones()

	for i in modification_physical_bones.physical_bone_chain_length:
		var p_bone_path := modification_physical_bones.get_physical_bone_node(i)
		var p_bone: PhysicalBone2D = _skeleton.get_node(p_bone_path)
		PhysicsServer2D.body_set_mode(p_bone.get_rid(), PhysicsServer2D.BODY_MODE_RIGID)
```

Whether or not the current implementation is considered experimental, the fix to this particular problem is simple and can be done right now. `PhysicalBone2D` should be passing these values via the inherited API and not access the server directly.

The usability of the whole thing is another matter, of course...

**Comment by Overkane** [comment]
Wasted several hours trying to get how to make physical bones to work until I found this issue. Perhaps PhysicalBone2D could be marked as "Experimental" so people will know it is not may work as expected?

**Comment by athyfr** [comment]
In reference to [@StrikeForceZero's workaround,](#issuecomment-1732113533) I notice some jitter on the Skeleton2D ragdoll that isn't present in the Rigidbody-based ragdoll. It is especially noticeable when the ragdoll reaches the ground, as it never stops twitching:

![Image](https://github.com/user-attachments/assets/9c99002b-4220-4355-bfec-1a471276b457)
(Left is Skeleton2D, right is regular Rigidbodies and Pinjoints)

Is this due to the way the Bone2Ds and PhysicalBone2Ds sync their positions?

**Comment by InTheProcess** [comment]
I discovered that Godot hates really small ragdolls.  Mine are small because I am making a pixelart game.  Think 32x32 pixels.

The pieces are barely held together, hands and feet clip through the floor, and it explodes when you apply an impulse to it.
